### DISCUSSION TITLE
miniShop2 3.0

### DISCUSSION THREAD

**Discussion by biz87** [Ibochkarev/miniShop2#515] [Ideas]
Примерный план работ для перехода на 3.0

**Защита от дурака**
Так как мы запланировали довольно большой фронт работ, часть из которых может сломать существующие проекты при бездумном обновлении нужно подумать над защитой. 
Как вариант выпуск отдельного пакета, и установка его поверх старого, чтобы просто обновиться было нельзя.  
Еще варианты?

**Нейминг**
Есть  три варианта

- Мы остаемся с неймингом minishop2 и выпускаем версию 3.0  этот по идее самый правильный, если выпускаем просто обновление
- Мы отказываемся от цифры 2 в названии и делаем версию miniShop 1.0 Этот вариант тоже имеет право на жизнь, если мы выпускаем отдельный пакет с защитой от дурака.  Но Здесь много потенциальных проблем с поддержкой сторонних пакетов
- Чисто для количества накидаю еще вариант MiniShop3 как логичное развитие miniShop2

**Вынос paypal в отдельный пакет**
В целом все понятно.  Вопросы. 
Кто сделает?  От чьего имени выкладывать?  Пакет будет платным или бесплатным?

**Миникорзина полноценная**
Нужно накидать схему компонента.  Что именно подразумевается под полноценной корзиной? 
Кто займется?   Опять же вроде есть такой компонент уже.  

**Обновление опций товара внутри корзины**
Пока такой возможности нет. 

**Переход фронта на jquery Free**
Вопросы:  Кто делает?  Какую версию ES будем использовать?  Традиционный ES5 или уже давно поддерживаемый
современными браузерами ES6?    Будем ли оглядываться на старые браузеры?

**Выпилить growl**
Продумать логику отдельно подключаемых библиотек
Определиться будем ли класть в коробку библиотеку?  

**Конструктор расширения моделей**
Добавление полей в любые таблицы минишопа по принципу  [msAddField](https://modx.pro/components/20947)

**Еще был разговор за отчеты и графики**
Выпилить графики, оставить все как есть или добавить какой то конструктор управления отчетами?
Если добавлять конструктор то с визуальным отображением или экспортом ?

**Comment by alroniks** [comment]
Пейпал могу взять себе. По хорошему, так как все платежки платные, его тоже стоит делать платным. Но с другой стороны, он всегда был из коробки и бесплатный, в теории можно так и оставить. Но я бы все же шел по пути Pro версии, куда платежки были бы включены по умолчанию. В зависимости от выбранного плана.

Миникорзина да, есть, но Марат к сожалению не сможет больше помочь с ее развитием. Пока что я попросил у Лёни дать мне пакет, мне он для работы был нужен, но я планировал некоторые пакеты Марата привести в порядок и вернуть в магазин.

Если хотите делать это из коробки, то это только плюс, можно за основу брать его компонент. Так даже лучше будет, так как не будет споров насчет платности/бесплатности компонента и его поддержки.

**Reply by Ibochkarev** [reply]
Мое мнение:
- paypall вынести и оставить бесплатно
- миникорзину обязательно делать из коробки

**Comment by alroniks** [comment]
Дополню мысль по поводу нейминга и про версии.

miniShop2 как основа может остаться, т.е. это базовый пакет, который в себя включает самое основное. Но для коммерции я бы делал Pro версию, вместо краудфандинга. Называть это miniShop2 Pro немного неправильно, так как немного мемно получается. Как бы мини, как бы 2 версию и еще прилепили про. Тут самое время придумать отдельное название (конкурс можно устроить :) ), и расписать толково тарифы, куда включить наборы функций. Причем я вижу так, что можно делать как вертикальные тарифы - чем больше функций, тем дороже, так и горизонтальные, когда решение затачивается под конкретную функцию, с некоторыми вариациями. Например, простой магазин, тариф 1, в него включен 1 пакет платежной системы и 1 пакет доставки, клиент на момент оплаты еще не знает, какой провайдер он будет использовать, но когда нужно, просто устанавливает себе нужный, счетчик лицензий обнуляется. И так же можно собирать решения по нишам - интернет-магазин детских товаров, ИМ товаров для дома, ИМ строительных товаров и прочее. Такие сборки могут тестироваться командой на предмет соответствия требованиям и могут закрывать потребности клиента прямо вот как есть. Останется только оформление подтянуть.

**Comment by Ibochkarev** [comment]
- Нейминг оставить как есть, только изменить на 3ку.
- PayPal вынести и сделать бесплатным
- внести в коробку полноценную миникорзину, за основу, как написал Иван взять компонент Марата
- переход на ES6
- убрать growl и как предлагал Баха сделать подключение нужных библиотек
- удобное добавление свой товара
- раздел со статистикой по продажам
- также посмотреть на конкурента Shopoholic - в разделе хороших практик

**Reply by CrazyBoy49z** [reply]
> * убрать growl и как предлагал Баха сделать подключение нужных библиотек

Cейчас это и так реализовано. Только по умолчанию jGrowl

**Reply by Sentinel-7** [reply]
@Ibochkarev 

> убрать growl

предлагаю все таки по умолчанию сделать SweetAlert2 хотя бы.. чтобы не было потока сообщений: "а где уведомлялки?"

**Reply by Ibochkarev** [reply]
@Sentinel-7 https://github.com/Ibochkarev/miniShop2/pull/542 уже реализовано

**Comment by Ruslan-Aleev** [comment]
Вопрос с неймингом (без учета платной версии "Pro"): допустим, выйдет MODX 3 и для него нужен отдельный пакет miniShop, как он будет называться, `miniShop3 v1.0.0`, а для текущей ветки MODX - `miniShop2 v3.0.0`? Если делать так, то можно запутаться, и не понятно как поисковики будут ранжировать доки.

**Reply by EMDM45** [reply]
miniShop3, miniShop2  -говорит лишь о том, что это несовместимые компоненты.
Можно, конечно, оставить просто miniShop v2, v3, v4 v5 и т.п. Но тогда при обновлении могут быть проблемы, поэтому лучше их делать в виде отдельных компонентов. Как было написано в посте "защита от дураков".

**Comment by Ibochkarev** [comment]
+ RestAPI

**Comment by sergant210** [comment]
Минишоп тянет зависимость pdoTools, в который встроен Fenom, с которым сниппеты минишопа не умеют работать в нормальном режиме шаблонизатора.  Оставлю здесь для подумать.

**Reply by EMDM45** [reply]
На мой взгляд, лучше уж зависимость от pdoTools, чем копипастить "велосипед" в ms2. Другое дело - настроить нормальную работу с этим pdoTools и Fenom.

**Comment by EMDM45** [comment]
Моё мнение такое.
Нужно сосредоточиться на разработке (улучшении) самого движка магазина. Оставить его бесплатным. Это проще для разработки и дальнейшей поддержке, чем городить всякие платные тарифы и прочее, усложняя проект, а как следствие и плодить разные "дыры".
Всё остальное - отдельные компоненты. Платные или бесплатные - решает разработчик. Задача лишь в том, чтобы эти компоненты безболезненно и безкостыльно можно было подключать.
Собственно так это и сделано сейчас вроде. Что в этом плохого?
Можно и PayPal вынести в отдельные бесплатный компонент, чтобы еще больше разгрузить сам движок MS.

**Reply by CrazyBoy49z** [reply]
Плохое предложение, так как минишоп устаревший

**Reply by EMDM45** [reply]
Какое предложение и при чем тут устаревший минишоп? Наверное ошиблись адресатом))

**Comment by CrazyBoy49z** [comment]
Предлагаю для minishop2 3.0 взять минимальную версию php 7.2(хотелось б php 8, но ещё долго ждать) так как там есть строгая типизация. Могу в свободное время начать работу над этой веткой (если все готовые использовать php 7.2).

**Comment by wfoojjaec** [comment]
На счёт нейминга - не принципиально. Хотя просто miniShop мне больше по душе.

По поводу функционала.
Конструктор для расширения моделей очень нужен. Кровь из носа. Практически каждый магазин тянет за собой такую задачу. И качественное решение из коробки обрадует большую половину разработчиков. Причём для расширения почти всех моделей. Товар, заказ, товар в заказе и т д.
Ещё хорошей идеей будет отказ от jQuery, хотя и это не так важно.
Fenom. Спорный момент. Было бы круто, если бы компонент не зависел ни от Fenom ни от pdoTools, ни от jQuery. Мечтать не вредно, ведь это упирается в тормоза самого парсера MODX.

Отдельно хотелось бы подметить серьёзные проблемы с производительностью. На мой взгляд как конечное коммерческое решение этот магазин очень слабоват.

1. Зависимость от pdoTools, и, как следствие, болезнь getPage. Все id передаются в IN условие, по которому уже идёт второй запрос с LIMIT. Не трудно догадаться, сколько RAM пожирает запрос, содержащий в себе около 100к id товаров в условии. В админке это происходит абсолютно так же.
2. Проблемы с GROUP BY. В msProducts по-умолчанию сделана группировка по id. Зачем, я пока только догадываюсь. Связи в таблицах modResource к msProductData один к одному... А производительность у сочетания ORDER BY и GROUP BY оставляет желать лучшего - если убрать GROUP BY, то скорость возрастает примерно в 10 раз. Не шутка. Если группировка всё-таки необходима (для превьюшек, картинок), то их вполне можно тащить одельным легковесным запросом, когда уже известна конечная выдача (после применения LIMIT). Это займёт миллисекунды вместо секунд.
3. Вся модель построена на сочетании modResource и msProductData через JOIN. Это, конечно, прикольно, но при использовании критериев из обеих таблиц (а такое происходит почти всегда, published = 1, price > 100 и т д) базе ничего не остаётся, кроме как на каждый запрос создавать временную таблицу, сортировать её, фильтровать, группировать (см. выше) и только потом отдавать результат. Эти проблемы не заметны вообще, пока у магазина нет особой посещаемости. Но при её появлении оказывается, что на сервере, где другие CMF/CMS чувствуют себя легко и непринуждённо, MODX еле-еле выплёвывает страницы с фильтрами за 5-10 секунд.
4. Опции товаров. Они сделаны ужасно, ничем не лучше TV в MODX. И то и другое следует хотя-бы переделать на EAV-модель, чтобы числа были в таблице с числами, строки в таблице со строками, а текст... ну вы поняли. Фильтрация по ним сделана через CAST. Т е база на каждый такой фильтр делает CAST по всей протяжённости значений в уже любимых временных таблицах, и каждое значение сверяет с фильтром. И если на сайте, скажем 10к товаров, и у каждого по 10 опций, то каждый фильтр сканирует 100к значений снова и снова. Да, у mSearch2 есть "прокладка" в виде самопального кеша запросов и индексации, да у БД есть свой собственный кеш, да полно других костылей, которыми это можно подпереть. Но это не отменяет того факта, что кеши пропадают периодически при синхронизации товаров и их редактировании, и всегда есть счастливый клиент. который попадёт на страницу без кеша.

Я не знаю, как эти вещи сделаны внутри Commerce у modmore. Возможно ещё хуже. Возможно в разы лучше.
Кроме того, я не уверен, действительно ли эти проблемы с производительностью настолько актуальны, ведь это miniShop, а не megaShop ^_^

**Comment by Alekseevich-psk** [comment]
Вопрос, а если делать MS3, куда внести все выше описанные предложения и прочее. Может есть смысл подумать над сторонней админ-панелью для менеджеров? Как я понимаю, поверх MS2 часто пишется "что-то другое"!?

**Reply by biz87** [reply]
А какое отношение сторонняя админ-панель имеет к минишопу? Она на то и сторонняя, что делается под бизнес-процессы компании, которые чаще всего отличаются от типовой сборки.  Да и дизайн делается под клиента.

**Reply by Alekseevich-psk** [reply]
Не знаю, для этого и спросил). Логика была такая, в чате обсуждалась идея вырезать минишоп из modx, я вот и задумался, что возможно сделать его обособленным со своей админкой, но при этом частью modx. 
технически я понятие не имею как это сделать.. ) 
Ну или просто принять во внимание, что не многие могут делать кастомные вещи поверх MS2.

**Reply by wfoojjaec** [reply]
Отделять от MODX - плохая идея. Для таких целей уже существует новый шопкипер. Ну и компоненты магазинов в других фреймворках, которые можно прикрутить, используя composer.
А вот сделать структуру по аналогии с одной из более старых версий шопкипера - не плохая идея. Вынести товары полностью в отдельную таблицу, продублировав modResource и поставив их маршрутизацию на плагин.

**Reply by Alekseevich-psk** [reply]
Благодарю за разъяснения)

**Reply by sergant210** [reply]
Поддерживаю эту идею, так как я сторонник подхода "админка только для админа".

**Reply by EMDM45** [reply]
Мы куда-то не в ту сторону уходим, как мне кажется. Нужно сосредоточиться на самом движке (т.е. набором функций). Отдельная админка - это уже скорее отдельный компонент. Пытаться всё впихнуть в одну коробку под названием "minishop"... на мой взгляд не совсем целесообразно. Потом еще и поддержкой всего этого...

**Reply by sergant210** [reply]
> Отделять от MODX - плохая идея.

Почему? Минишоп - одна из главных причин использования MODX в RU сообществе. Так как перспективы MODX  сомнений не вызывают, то можно было бы иметь продукт, который не утонет вместе с ним. Многие разработчики к нему уже привыкли, тучу собак съели и могли бы и дальше на нём зарабатывать. Тем более в сообществе текучка. Разработчик ушел, но минишоп не бросает. Так думаю.

**Reply by Sentinel-7** [reply]
>     одна из главных причин использования MODX

я бы не сказал так.. очень много сайтов делается без ms2
тут больше идет уклон на удобстве использования в backend и кастомизации

**Reply by sergant210** [reply]
> я бы не сказал так.. очень много сайтов делается без ms2
MODX уже не на слуху. Рекламы нет. Сарафанное радио тихое и медленное. Студии его практически не используют. Курсов по нему нет. MODX - прерогатива фрилансеров. Как то обсуждали тему, что на  MODX выходят когда гуглят интернет-магазин.

**Reply by Sentinel-7** [reply]
Сколько людей, столько и мнений. У меня другая статистика... :wink:

**Reply by sergant210** [reply]
> Сколько людей, столько и мнений. У меня другая статистика... 😉

Мнение - это одно. Статистика - другое. Мнение может меняться в зависимости от фактов. Если по твоим данным спрос на MODX вырос, приму во внимание. Я только с ваших слов говорю ибо я не сайтостроитель. У меня нет полной статистики.

**Comment by ru2fac** [comment]
Задался тут целью передащить небольшой магазинчик на Modx revo 3. Развернул advanced сборку, поставил допы sdstore и оттуда же pdoTools. Следом ставлю minishop2-3.0.4-p12 и тут начинается непонятная штука.
Установка зависает на моменте:
``
Консоль запущена...
Пробуем установить пакет с подписью: minishop2-3.0.4-pl2
Пакет найден... сейчас идёт подготовка к его установке.
Загрузка рабочего пространства пакета...
Рабочее пространство загружено, сейчас устанавливаем пакет...
`

Дальне ничего не происходит, сколько бы не ждал. Пробую перезагрузить страницу, в списке менеджера пакетов ms2 по прежнему находится. как не установленный пакет, с активной кнопкой УСТАНОВИТЬ, пробую установить снова, лог уже длиннее:
Первая часть лога все ок, как обычно, затем куча ошибок с тем, что не может скопировать файл или создать каталог.Вот пример этой ошибки:
`xPDOCacheManager->copyFile(): Could not copy file /home/public_html/core/packages/minishop2-3.0.4-pl2/modCategory/5c48c0021bfaf9dc0cf5a6b1fbdaeccf/0/minishop2/img/mgr/extensions/pdf.png to /home/public_html/assets/components/minishop2/img/mgr/extensions/pdf.png
Could not copy file /home/public_html/core/packages/minishop2-3.0.4-pl2/modCategory/5c48c0021bfaf9dc0cf5a6b1fbdaeccf/0/minishop2/img/mgr/extensions/pdf.png to /home/public_html/assets/components/minishop2/img/mgr/extensions/pdf.png; could not create directory.`

Хостинг Beget, ModxRevo 3.0.1, minishop 3.0.4 PHP 7.4
Пробовал играть с версиями PHP, кодировкой MySQL(utf8_general_ci) тоже не помогло.

Как это можно пофиксить? Спасибо.

**Reply by webinmd** [reply]
minishop 3.0.4 не подготовлен к работе  на ModxRevo 3.0.1.
только для Modx < 3

**Reply by ru2fac** [reply]
Спасибо за скорый ответ. Буду ждать update компонента. Разработчикам респект.

**Comment by block888dev** [comment]
друзья как дела с 3?!

**Reply by biz87** [reply]
https://github.com/modx-pro/MiniShop3

**Reply by block888dev** [reply]
> https://github.com/modx-pro/MiniShop3

super!! апдейтнуться с 2 можно как-то??

**Reply by block888dev** [reply]
Привет, есть новости как надо с 2 перезжать на 3 в modx 3.0.3 на php8?
